src/Processors/Formats/Impl/AvroRowInputFormat.cpp

```diff
@@ -438,6 +434,39 @@ AvroDeserializer::DeserializeFn AvroDeserializer::createDeserializeFn(const avro
                         target_type->getName());
                 }
             }
+
+            if (target.isVariant())
+            {
+                const DataTypeVariant & variant_type = assert_cast<const DataTypeVariant &>(*target_type);
+                const auto & nested_types = variant_type.getVariants();
+
+                if (root_node->leaves() != nested_types.size())
+                    throw Exception(ErrorCodes::INCORRECT_DATA, "The number of union types in Avro record ({}) doesn't match the number of types in destination Variant type ({}).", root_node->leaves(), nested_types.size());
+
+                std::vector<DeserializeFn> nested_deserializers;
+                nested_deserializers.reserve(root_node->leaves());
+
+                for (int i = 0; i != static_cast<int>(root_node->leaves()); ++i)
+                {
+                    auto nested_deserializer = createDeserializeFn(root_node->leafAt(i), nested_types.at(i));
+                    nested_deserializers.emplace_back(nested_deserializer);
+                }
+
+                return [nested_deserializers](IColumn & column, avro::Decoder & decoder)
```

Comment:
Just out of curiosity, what happens under the hood if we don't move it? I thought that returning from a function implicitly discards all variables in the function scope. Or is it the "better safe than sorry" scenario when the compiler might copy the whole vector instead of moving it by itself?

Reply:
I am afraid that it is going to copy the vector to the closure during its creation. Maybe optimiser will understand this,  but in my opinion it is better to handle memory issues in a more verbose manner